### DISCUSSION TITLE
coupling le9db with zram

### DISCUSSION THREAD

**Discussion by gottaeat** [hakavlad/le9-patch#9] [Q&A]
i have a laptop w/ a core2 duo t8100 + ~1800mb of ram (~200mb memmap'd away for reasons) that i build heavy stuff with, which as you can imagine ends up with me swapping to disk and eventually getting livelocked. seeing this project made me think of coupling le9db with a 4gb lz4hc compressed zram as the only swap. my current setup is as follows:
```sh
$ swapon
NAME       TYPE      SIZE  USED PRIO
/dev/zram0 partition   4G 36.5M  100

MemTotal                           : 1838564 kB

/sys/block/zram0/disksize          : 4G
/sys/block/zram0/mem_limit         : 0

/proc/sys/vm/clean_min_kbytes      : 0
/proc/sys/vm/clean_low_kbytes      : 524288

/proc/sys/vm/swappiness            : 200
/proc/sys/vm/vfs_cache_pressure    : 200
/proc/sys/vm/watermark_scale_factor: 200
```
the system with these options allows for browsing the web while building the rust toolchain with two jobs and thinlto, so i guess i am on the right track. what i would like to ask is whether these values are sane or if they could be improved.

thank you.

**Comment by hakavlad** [comment]
>lz4hc

Why not zstd?

>/proc/sys/vm/vfs_cache_pressure    : 200
>/proc/sys/vm/watermark_scale_factor: 200

Why did you choose these values? 

>/proc/sys/vm/clean_low_kbytes      : 524288

Perhaps this is too big a value for ~1800mb of ram. 

Maybe setting clean_low > 0 is not necessary if you set swappiness=200.

I'd also like to recommend [nohang](https://github.com/hakavlad/nohang) with `zram_checking_enabled=True` config option.

**Reply by gottaeat** [reply]
> why not zstd?

wanted the harshest compression available, can try zstd but with lz4hz i saw 3.1G swap usage with rust so not sure.

>Why did you choose these values?

`vfs_cache_pressure` is set to 200 to hold the minimum amount of cache in the memory and i saw `watermark_scale_factor being` being set to 200 in one of the tests for swapping large chunks of data at a time.

> Maybe setting clean_low > 0 is not necessary if you set swappiness=200.

i will try that.

i appreciate your input, my point was to sort of make a huge zram and use that as the swap and tell the kernel to swap as much as possible as i am okay with the performance hit of clearing cache and swapping all the time. do you have any extra input considering all of the sys/procfs options in the equation?

**Comment by owl4ce** [comment]
**My setup.** No problems to build **Beyond Linux from Scratch** and **Gentoo** (multitask, core/threads fully used).
**Specs:** AMD A4-5000 with 4G (3.3G) RAM (+ zRAM with size 3x of RAM)

<p align="center"><img src="https://user-images.githubusercontent.com/53987136/127678296-08f5ce98-5741-4113-98e9-511d44f06fa5.png"/></p>

Building rust and LLVM with ThinLTO usually use around 2-4G of my zRAM. I prefer lz4 for performance, using high compression algorithm may overhead the CPU and I/O utilization.

**Reply by hakavlad** [reply]
why did you choose swappiness=10?

**Reply by owl4ce** [reply]
> why did you choose swappiness=10?

To prevent overloaded CPU usage if it's doing heavy task concurrently with swapping memory pages (especially if it's compressed). I'm still relying on direct usage in RAM instead of using Swap from zRAM completely.

> _zRAM only used when my RAM is running out of space._

Remember that chunks in the swap not cleared after bringing back to physical memory. In the RAM, you have the **m** pages in **zram** and the **n** pages that are back in **memory**. When the application go back in the background, the kernel doesn't write back those pages into swap. So your RAM usage getting bigger over time every swapped, although zram is dynamically allocated.

Correct me if I'm wrong.
